src/tracker/daemon.rs

```diff
@@ -0,0 +1,153 @@
+          resp.extend_from_slice(&self.peer_list(&req.infohash));
+          self.sock.send_to(&resp, peer).unwrap();
+
+          self.insert(req.infohash, peer);
+        }
+      }
+    }
+  }
+
+  pub fn insert(&mut self, infohash: [u8; 20], addr: SocketAddr) {
+    if let Some(set) = self.records.get_mut(&infohash) {
+      set.insert(addr);
+    } else {
+      let mut set = HashSet::new();
+      set.insert(addr);
+      self.records.insert(infohash, set);
+    }
+  }
+
+  fn peer_list(&self, infohash: &[u8; 20]) -> Vec<u8> {
+    match self.records.get(infohash) {
+      None => vec![],
+      Some(set) => Self::compact_peer_list(set),
+    }
+  }
+
+  fn compact_peer_list(set: &HashSet<SocketAddr>) -> Vec<u8> {
+    let mut v = Vec::new();
+    for p in set {
+      match p.ip() {
+        IpAddr::V4(ip) => v.extend_from_slice(&ip.octets()),
+        IpAddr::V6(ip) => v.extend_from_slice(&ip.octets()),
+      }
+      v.extend_from_slice(&p.port().to_be_bytes());
+    }
+    v
+  }
+}
+
+mod tests {
+  use super::*;
+
+  #[test]
+  fn run() {
+    let (_, addr) = Daemon::spawn();
+
+    let mut c = Client::connect(addr).unwrap();
+    let mut a = c.local_addr();
+    let mut resp = c.announce_exchange(&[0u8; 20].into()).unwrap();
+
+    for i in 0..4 {
+      assert_eq!(resp.len(), i);
+      c = Client::connect(addr).unwrap();
+      resp = c.announce_exchange(&[0u8; 20].into()).unwrap();
+      assert!(resp.contains(&a));
+      a = c.local_addr();
+    }
+  }
+
+  #[test]
+  fn separate_infohashes() {
+    let (_, addr) = Daemon::spawn();
+    let c1 = Client::connect(addr).unwrap();
+    let c2 = Client::connect(addr).unwrap();
+    let a1 = c1.local_addr();
+    let a2 = c2.local_addr();
+
+    let infohash1 = Infohash::from(rand::thread_rng().gen::<[u8; 20]>());
+    let infohash2 = Infohash::from(rand::thread_rng().gen::<[u8; 20]>());
+    let resp1 = c1.announce_exchange(&infohash1).unwrap();
+    let resp2 = c2.announce_exchange(&infohash2).unwrap();
+    assert_eq!(resp2.len(), 0);
+    assert_eq!(resp1.len(), 0);
+
+    let resp1 = c1.announce_exchange(&infohash1).unwrap();
+    let resp2 = c2.announce_exchange(&infohash2).unwrap();
+    assert_eq!(resp1.len(), 1);
+    assert_eq!(resp2.len(), 1);
+    assert!(resp1.contains(&a1));
+    assert!(resp2.contains(&a2));
+  }
+
+  #[test]
+  fn reannounce() {
+    let (_, addr) = Daemon::spawn();
+    let infohash = Infohash::from(rand::thread_rng().gen::<[u8; 20]>());
+    let c1 = Client::connect(addr).unwrap();
+    let c2 = Client::connect(addr).unwrap();
+    let a1 = c1.local_addr();
+    let a2 = c2.local_addr();
+    let resp1 = c1.announce_exchange(&infohash).unwrap();
+    let resp2 = c2.announce_exchange(&infohash).unwrap();
+    let resp3 = c1.announce_exchange(&infohash).unwrap();
+
+    assert_eq!(resp1.len(), 0);
+    assert_eq!(resp2.len(), 1);
+    assert_eq!(resp3.len(), 2);
+    assert!(resp3.contains(&a1));
+    assert!(resp3.contains(&a2));
+  }
```

Comment:
Missing test: create a torrent file, generate a magnet link, seed the info dict, fetch the info dict, and recreate the torrent file from the info dict.